wren-core-py/src/extractor.rs

```diff
@@ -0,0 +1,326 @@
+                mdl: WrenMDL::new_ref(manifest),
+            })
+    }
+
+    /// parse the given SQL and return the list of used table name.
+    pub fn resolve_used_table_names(&self, sql: &str) -> Result<Vec<String>, CoreError> {
+        resolve_used_table_names(&self.mdl, sql)
+    }
+
+    /// Given a used dataset list, extract manifest by removing unused datasets.
+    /// If a model is related to another dataset, both datasets will be kept.
+    /// The relationship between of them will be kept as well.
+    /// A dataset could be model, view.
+    pub fn extract_by(
+        &self,
+        used_datasets: Vec<String>,
+    ) -> Result<PyManifest, CoreError> {
+        extract_manifest(&self.mdl, &used_datasets)
+    }
+}
+
+fn resolve_used_table_names(mdl: &WrenMDL, sql: &str) -> Result<Vec<String>, CoreError> {
+    let ctx_state = wren_core::SessionContext::new().state();
+    ctx_state
+        .sql_to_statement(sql, "generic")
+        .map_err(CoreError::from)
+        .and_then(|stmt| {
+            ctx_state
+                .resolve_table_references(&stmt)
+                .map_err(CoreError::from)
+        })
+        .map(|tables| {
+            tables
+                .iter()
+                .filter(|t| {
+                    t.catalog().map_or(true, |catalog| catalog == mdl.catalog())
+                        && t.schema().map_or(true, |schema| schema == mdl.schema())
+                })
+                .map(|t| t.table().to_string())
+                .collect()
+        })
+}
+
+fn extract_manifest(
+    mdl: &WrenMDL,
+    used_datasets: &[String],
+) -> Result<PyManifest, CoreError> {
+    let extracted_models = extract_models(mdl, used_datasets);
+    let (used_views, models_of_views) = extract_views(mdl, used_datasets);
+    let used_models = [extracted_models, models_of_views].concat();
+    let used_relationships = extract_relationships(mdl, &used_models);
+    Ok(PyManifest {
+        catalog: mdl.catalog().to_string(),
+        schema: mdl.schema().to_string(),
+        models: used_models,
+        relationships: used_relationships,
+        metrics: mdl.metrics().to_vec(),
+        views: used_views,
+    })
+}
+
+fn extract_models(mdl: &WrenMDL, used_datasets: &[String]) -> Vec<Arc<Model>> {
+    let mut used_set: HashSet<String> = used_datasets.iter().cloned().collect();
+    let mut stack: Vec<String> = used_datasets.to_vec();
+    while let Some(dataset_name) = stack.pop() {
+        if let Some(model) = mdl.get_model(&dataset_name) {
+            model
+                .columns
+                .iter()
+                .filter_map(|col| {
+                    col.relationship
+                        .as_ref()
+                        .and_then(|rel_name| mdl.get_relationship(rel_name))
+                })
+                .flat_map(|rel| rel.models.clone())
+                .filter(|related| used_set.insert(related.clone()))
+                .for_each(|related| stack.push(related));
```

Comment:
We can change`used_set` to a HashMap
```rust
    let mut used_set: HashMap<String, usize> = used_datasets.iter().map(|s| (s.clone(), 0)).collect();
```
Then, we can use `Entry` to avoid cloning `related` if it has existed.
```suggestion
                .for_each(|related| {
                    if let Entry::Vacant(vacant) = used_set.entry(related) {
                        let key = vacant.key().clone();
                        vacant.insert(0);
                        stack.push(key);
                    }
                });
```

Reply:
If we care about clone stack, we can use `HashSet:: contains`.
```suggestion
                .for_each(|related| {
                    if !used_set.contains(&related) {
                        used_set.insert(related.clone());
                        stack.push(related);
                    }
                });
```